server/vmap/trillian_map_server.go

```diff
@@ -74,24 +74,33 @@ func (t *TrillianMapServer) GetLeaves(ctx context.Context, req *trillian.GetMapL
 
 	smtReader := merkle.NewSparseMerkleTreeReader(req.Revision, hasher, tx)
 
-	leaves, err := tx.Get(ctx, req.Revision, req.Index)
-	if err != nil {
-		return nil, err
-	}
-	glog.Infof("%v: wanted %v leaves, found %v", mapID, len(req.Index), len(leaves))
+	// TODO: we need to return one response per request.
+	// Get only returns items for things that exist.
 
-	resp := &trillian.GetMapLeavesResponse{
-		MapLeafInclusion: make([]*trillian.MapLeafInclusion, len(leaves)),
-	}
-	for i, leaf := range leaves {
-		proof, err := smtReader.InclusionProof(ctx, req.Revision, leaf.Index)
+	inclusions := make([]*trillian.MapLeafInclusion, 0, len(req.Index))
+	for i, index := range req.Index {
+		proof, err := smtReader.InclusionProof(ctx, req.Revision, index)
 		if err != nil {
 			return nil, err
 		}
-		// Copy the leaf from the iterator, which gets overwritten
-		value := leaf
+
+		// Create placeholder leaf for non-existant leaves.
+		leaf := &trillian.MerkleLeaf{
+			Index: index,
+		}
+
+		// Fetch the leaf if it exists.
+		leafs, err := tx.Get(ctx, req.Revision, [][]byte{index})
+		if err != nil {
+			return nil, err
+		}
+		for l := range leafs {
```

Comment:
nit: I would prefer an if to a for here, so it's clearer we expect/want at most a size 1 slice.

```go
if len(leaves) == 1 {
  leaf = leaves[0]
}
```